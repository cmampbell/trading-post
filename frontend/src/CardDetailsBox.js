import React, { useState } from "react";
import { Box, FormControl, InputLabel, Select, MenuItem, Button, Typography, TextField } from "@mui/material";

// TODO: Add quantity form field
const CardDetailsBox = ({ cards, setListCards, handleClose }) => {
    // rendered after user selects a card
    const INITIAL_STATE = { card: cards[0], condition: 'Lightly Played', foil: 'No', quantity: 1 }
    // form control
    const [cardDetails, setCardDetails] = useState(INITIAL_STATE)

    // update state based on input changes
    const handleChange = (evt) => {
        const { name, value } = (evt.target)
        if (name === 'set') {
            const [setName, collector_number] = evt.target.value.split('-');
            const newCard = cards.find((card) => card.set_name === setName && card.collector_number === collector_number)
            setCardDetails(() => ({ ...cardDetails, card: { ...newCard } }))
        } else {
            setCardDetails(() => ({ ...cardDetails, [name]: value }))
        }
    }

    // add card object to list on click
    const handleClick = (evt) => {
        evt.preventDefault();
        const {card, condition, foil} = cardDetails;
        setListCards((currentCardList)=> [...currentCardList, {condition, foil, ...card}])
        setCardDetails(()=> {})
        handleClose();
    }

    return (
        <Box sx={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'flex-start',
            width: '90%',
            height: '100%',
            m: 2
        }}>
            {/* container for image */}
            <Box sx={{
                height: '30%',
                overflow: 'hidden',
            }}>
                <img
                    src={cardDetails.card.art_uri}
                    alt={`${cardDetails.card.name} image`}
                    style={{ width: 'auto', height: '100%' }}
                />
            </Box>

            <h3>{cardDetails.card.name}</h3>

            <Box sx={{
                display: 'flex',
                flexDirection: 'row',
                flexWrap: 'wrap',
                alignItems: 'center',
                justifyContent: 'space-between',
                alignContent: 'flex-start',
                width: '80%',
                height: 'auto'
            }}>
                <FormControl
                    sx={{ m: 1, width: '90%' }}
                    size="small"
                >
                    <InputLabel id="set-select-label">Set</InputLabel>
                    <Select
                        labelId="set-select-label"
                        id="set-select"
                        value={`${cardDetails.card.set_name}-${cardDetails.card.collector_number}`}
                        label="set"
                        name="set"
                        onChange={handleChange}
                    >
                        {/* render menu items for each version of the card found */}
                        {cards.map((card) =>
                            <MenuItem
                                value={`${card.set_name}-${card.collector_number}`}
                                key={`${card.set_code}-${card.collector_number}`}
                            >{card.set_name} - #{card.collector_number}</MenuItem>
                        )}
                    </Select>
                </FormControl>
                <FormControl
                    sx={{ m: 1, width: '90%' }}
                    size="small"
                >
                    <InputLabel id="condition-select-label">Condition</InputLabel>
                    <Select
                        labelId="condition-select-label"
                        id="condition-select"
                        value={cardDetails.condition}
                        label="condition"
                        onChange={handleChange}
                        name="condition"
                    >
                        {/* using condition list from tcgplayer.com */}
                        <MenuItem value="Near Mint" key="mint">Near Mint</MenuItem>
                        <MenuItem value="Lightly Played" key="lightly-played">Lightly Played</MenuItem>
                        <MenuItem value="Moderately Played" key="moderately-played">Moderately Played</MenuItem>
                        <MenuItem value="Heavily Played" key="heavily-played">Heavily Played</MenuItem>
                        <MenuItem value="Damaged" key="damaged">Damaged</MenuItem>
                    </Select>
                </FormControl>
                <FormControl
                    sx={{ m: 1, width: '90%' }}
                    size="small"
                >
                    <InputLabel id="foil-select-label">Foil</InputLabel>
                    <Select
                        labelId="foil-select-label"
                        id="foil-select"
                        value={cardDetails.foil}
                        label="foil"
                        name="foil"
                        onChange={handleChange}
                    >
                        {/* certain cards only have certain prices, only render the ones they have */}
                        {cardDetails.card.usd_price && <MenuItem value="No" key="no">No</MenuItem>}
                        {cardDetails.card.usd_foil_price && <MenuItem value="Yes" key="yes">Yes</MenuItem>}
                        {cardDetails.card.usd_etched_price && <MenuItem value="Etched" key="etched">Etched</MenuItem>}
                    </Select>
                </FormControl>
                {/* TO-DO: add quantity */}
                {/* <FormControl
                    sx={{ m: 1, width: '90%' }}
                    size="small"
                >
                    <InputLabel id="quantity-select-label">Foil</InputLabel>
                    <TextField inputProps={{ inputMode: 'numeric', pattern: '[0-9]*' }} />
                </FormControl> */}
            </Box>
            <Typography sx={{ m: 1 }}>
                {/* depending on the foil selection, render relevant price */}
                ${cardDetails.foil === "Yes" ? cardDetails.card.usd_foil_price
                    : cardDetails.foil === "Etched" ? cardDetails.card.usd_etched_price
                        : cardDetails.card.usd_price}
            </Typography>
            <Button onClick={handleClick}> Add Card! </Button>
        </Box>
    )
}

export default CardDetailsBox;