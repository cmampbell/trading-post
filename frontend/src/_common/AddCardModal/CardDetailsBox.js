import React, { useState } from "react";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import CardForm from "../CardForm/CardForm";

/* Returns MUI box component with several input boxes and a button.
*  
*  Rendered in AddCardModal after user has selected a card using <Searchbar/>
*
*  state:
*      card - the active card selected from printings.
*             Initially the first card object element in printings.
*
*  props:
*        printings - printings of card selected with <Searchbar />, returned from API
*        addCard - function used to add card to relevant memory.
*                State for trade page.
*                db for <CardBinder/> pages.
*        handleClose - function to close modal on form submit
*        fields - fields for <CardForm/>
*
*  Displays the name and artwork from the selected card, so user can see visual
*  confirmation that card art from db matches their physical card.
*
*  The artwork will have a black border for non-foil cards, and a rainbow border
*  for foil cards. User can change foil type in <CardForm/>
*
*  Renders <CardForm/>, which requires fields as a prop.
*
*  <CardDetailsBox /> should only care about which card is selected.
*/
const CardDetailsBox = ({ printings, addCard, handleClose, fields }) => {

    const [card, setCard] = useState(printings[0]);

    const borderStyles = card.foil === 'Yes' || card.foil === 'Etched' ?
        {
            border: '5px solid transparent',
            borderImage: 'linear-gradient(to bottom right, #b827fc 0%, #2c90fc 25%, #b8fd33 50%, #fec837 75%, #fd1892 100%)',
            borderImageSlice: 1,
        }
        : {
            border: 5,
            borderColor: 'black',
            borderRadius: '10px'
        };

    return (
            <Box sx=
                {{
                    display: 'flex',
                    flexDirection: 'column',
                    alignItems: 'center',
                    justifyContent: 'flex-start',
                    marginTop: 4,
                }}
            >
                {/* container for image */}
                <Box sx={{
                    height: 250,
                    width: 'auto',
                    overflow: 'hidden',
                    ...borderStyles
                }}>
                    <img
                        src={card.art_uri}
                        alt={`${card.name} art by ${card.artist}`}
                        style={{
                            width: 'auto',
                            height: '100%',
                         }}
                    />
                </Box>
                <Typography variant='subtitle2' sx={{ m: 1, fontSize: '11px' }}>
                    Illustrated by {card.artist}. &#8482; & &copy; Wizards Of The Coast, Inc.
                </Typography>

                <Typography variant='h6'>{card.name}</Typography>
                {/* TO-DO: Add context provider here */}
                <CardForm
                    card={card}
                    setCard={setCard}
                    printings={printings}
                    handleClose={handleClose}
                    fields={fields}
                    addCard={addCard}
                    type={'Add'}
                />
            </Box>
    )
}

export default CardDetailsBox;