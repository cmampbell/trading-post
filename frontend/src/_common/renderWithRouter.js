import React, { isValidElement } from "react";
import { render } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";


// borrowed from: https://webup.org/blog/how-to-avoid-mocking-in-react-router-v6-tests/

export default renderWithRouter = (children, routes=[]) => {
    // options creating routes for valid children
    // else keep children as they are
    // first argument is a full route

    const options = isValidElement(children)
            ? { element: children, path: '/'}
            : children;

    // spread out the options here
    // the options are a route object
    // then routes are spread and passed after
    const router = createMemoryRouter([{...options}, ...routes],
        {
        initialEntries: [options.path],
        initialIndex: 1
    }
    );

    return render(<RouterProvider router={router}/>);
}