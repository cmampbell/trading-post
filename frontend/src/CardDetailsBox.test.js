import React from "react";
import renderWithRouter from "./renderWithRouter";
import CardDetailsBox from "./CardDetailsBox";
import userEvent from '@testing-library/user-event'
import { act } from "react-dom/test-utils";
import { waitFor, screen, queryByLabelText } from "@testing-library/react";

const testCard = { name: 'Island', image_uri: 'test_uri', usd_price: '7.65', id: '1' , setCode: 'IXL'}
const testCard2 = { name: 'Mountain', image_uri: 'test_uri', usd_price: '7.65', id: '2' , setCode: 'IXL'}

describe('CardList tests', () => {
    it('should render without crashing', () => {
        renderWithRouter(<CardDetailsBox cards={[testCard, testCard2]} />);
    })

    it('should match snapshot', () => {
        const { asFragment } = renderWithRouter(<CardDetailsBox cards={[testCard, testCard2]} />);

        expect(asFragment()).toMatchSnapshot();
    })

    it('should render all necessary fields', async () => {
        const { queryByText, queryByAltText, queryByLabelText } = renderWithRouter(<CardDetailsBox cards={[testCard, testCard2]}/>);

        expect(queryByText('Island')).toBeInTheDocument();
        expect(queryByAltText('Island image')).toBeInTheDocument();
        expect(queryByLabelText('Set')).toBeInTheDocument();
        expect(queryByLabelText('Condition')).toBeInTheDocument();
        expect(queryByLabelText('Foil')).toBeInTheDocument();
        expect(queryByText(`$${testCard.usd_price}`)).toBeInTheDocument();
        expect(queryByText(`Add Card!`)).toBeInTheDocument();
    })
})