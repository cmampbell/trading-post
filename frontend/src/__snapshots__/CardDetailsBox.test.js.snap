// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1mz8pzr"
  >
    <div
      class="MuiBox-root css-1cbu74j"
    >
      <img
        alt="Island image"
        style="width: auto; height: 100%;"
      />
    </div>
    <h3>
      Island
    </h3>
    <div
      class="MuiBox-root css-becfju"
    >
      <div
        class="MuiFormControl-root css-ssswi9-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="set-select-label"
        >
          Set
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="set-select-label set-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="set-select"
            role="button"
            tabindex="0"
          >
             - #
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="set"
            tabindex="-1"
            value="undefined-undefined"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                set
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root css-ssswi9-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="condition-select-label"
        >
          Condition
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="condition-select-label condition-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="condition-select"
            role="button"
            tabindex="0"
          >
            Lightly Played
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="condition"
            tabindex="-1"
            value="Lightly Played"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                condition
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root css-ssswi9-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="foil-select-label"
        >
          Foil
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="foil-select-label foil-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="foil-select"
            role="button"
            tabindex="0"
          >
            No
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="foil"
            tabindex="-1"
            value="No"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                foil
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-86ygas-MuiTypography-root"
    >
      $7.65
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
       Add Card! 
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
